### DISCUSSION TITLE
Docs: Little ambiguity regarding path operation vs sub-dependencies "def function" running in threadpool

### DISCUSSION THREAD

**Discussion by srikarplus** [fastapi/fastapi#9912] [Questions] [answered]
### Description

Hi all, I'm going through the fastapi documentation, and I came across a little ambiguity. Requesting your help to resolve my ambiguity.

At https://fastapi.tiangolo.com/async/#path-operation-functions, it was mentioned that

> When you declare a path operation function with normal **def** instead of async def, **it is run in an external threadpool that is then awaited**, instead of being called directly (as it would block the server).

and at https://fastapi.tiangolo.com/async/#sub-dependencies

> the ones created with normal **def** would be called on an external thread (from the threadpool) instead of being "awaited".

At one place it was said that the `def function` were run in external threadpool and "**is awaited**" while at the second place it's said that "**not awaited**".

Can someone help me with the above confusion.

### FastAPI Version

0.100.0

### Python Version

3.10

**Comment by jgould22** [comment]
The first one is _awaiting the request to the thread pool_  in pseudo code something like 
```
def path_op_func(params):
   ...

result = await pool.submit(path_op_func(params))
```

the second is saying a similar thing but it is saying instead of awaiting `path_op_func` we submit it to a thread pool and await on that (because you wouldnt await a non async function)

**Reply by srikarplus** [reply]
hi @jgould22, thanks for taking time to clear my doubts. 
I understood the first one from your example but I'm still unable to grasp the second point...specifically `we submit it to a thread pool and await on that`. How is it different from first example where we are submitting it to pool and still awaiting.

**Reply by jgould22** [reply, marked as answer]
Ok let me try again

For the first

> When you declare a path operation function with normal def instead of async def, 
```python
# like this
def my_path_func():
```
>it is run in an external threadpool that is then awaited,
```python
# we do this
result = await thread_pool.submit(my_path_func())
```
 > **instead** of being called directly (as it would block the server).
```python
# NOT this because this blocks the server from answering requests
result = my_path_func()
```

for the second one

> the ones created with normal def 
```python
# like this
def my_path_func():
```
> would be called on an external thread (from the threadpool) 
```python
# we do this
result = await thread_pool.submit(my_path_func())
```
>instead of being "awaited".
```python
# because we cannot do this
result = await my_path_func()
# because my_path_func() is not a co-routine/task
```

The fact that you cannot call `await my_path_func()` because it is a normal function and not a task is implied so we send it to the thread pool.

Each snippet is essentially saying the same thing the second one is simply making the assumption that you understand already that you cannot await a non coroutine / task
